openmls/src/group/tests_and_kats/tests/remove_operation.rs

```diff
@@ -4,6 +4,132 @@ use crate::group::tests_and_kats::utils::{generate_credential_with_key, generate
+        bob_credential_with_key_and_signer.clone(),
+    );
+    let charlie_key_package = generate_key_package(
+        ciphersuite,
+        Extensions::empty(),
+        charlie_provider,
+        charlie_credential_with_key_and_signer,
+    );
+
+    // Define the MlsGroup configuration
+    let mls_group_create_config = MlsGroupCreateConfig::builder()
+        .ciphersuite(ciphersuite)
+        .build();
+
+    // === Alice creates a group ===
+    let mut alice_group = MlsGroup::new_with_group_id(
+        alice_provider,
+        &alice_credential_with_key_and_signer.signer,
+        &mls_group_create_config,
+        group_id,
+        alice_credential_with_key_and_signer
+            .credential_with_key
+            .clone(),
+    )
+    .expect("An unexpected error occurred.");
+
+    // === Alice adds Bob & Charlie ===
+
+    let (_message, welcome, _group_info) = alice_group
+        .add_members(
+            alice_provider,
+            &alice_credential_with_key_and_signer.signer,
+            &[
+                bob_key_package.key_package().clone(),
+                charlie_key_package.key_package().clone(),
+            ],
+        )
+        .expect("An unexpected error occurred.");
+    alice_group
+        .merge_pending_commit(alice_provider)
+        .expect("error merging pending commit");
+
+    let welcome: MlsMessageIn = welcome.into();
+    let welcome = welcome
+        .into_welcome()
+        .expect("expected message to be a welcome");
+
+    let mut bob_group = StagedWelcome::new_from_welcome(
+        bob_provider,
+        mls_group_create_config.join_config(),
+        welcome.clone(),
+        Some(alice_group.export_ratchet_tree().into()),
+    )
+    .expect("Error creating staged join from Welcome")
+    .into_group(bob_provider)
+    .expect("Error creating group from staged join");
+
+    let mut charlie_group = StagedWelcome::new_from_welcome(
+        charlie_provider,
+        mls_group_create_config.join_config(),
+        welcome,
+        Some(alice_group.export_ratchet_tree().into()),
+    )
+    .expect("Error creating staged join from Welcome")
+    .into_group(charlie_provider)
+    .expect("Error creating group from staged join");
+
+    // === Remove operation ===
+
+    let bob_index = bob_group.own_leaf_index();
+
+    alice_group
+        .remove_members(
+            alice_provider,
+            &alice_credential_with_key_and_signer.signer,
+            &[bob_index],
+        )
+        .expect("error removing member the first time");
+
+    alice_group.merge_pending_commit(alice_provider).unwrap();
+
+    // try removing bob a second time
+    alice_group
+        .remove_members(
+            alice_provider,
+            &alice_credential_with_key_and_signer.signer,
+            &[bob_index],
+        )
+        .expect_err("using API to re-remove someone should fail");
+
+    // TODO: craft another remove commit that removes bob's (blank) index and check that merging
```

Comment:
How do we track these tests? Do you want to add it on the dashboard or an issue?

Reply:
I would track it on the dashboard. I think tracking it in individual issues would be too messy, and creating a single tracking github issue just means we re-do the work of the dashboard. Unless we automate it :nerd_face:

Reply:
https://github.com/openmls/validation.dhall/pull/14